core/src/main/java/org/infinispan/reactive/publisher/impl/LocalPublisherManagerImpl.java

```diff
@@ -169,70 +284,116 @@ public void segmentsLost(IntSet lostSegments) {
+   }
 
-                  CompletionStage<R> stage = transformer.apply(innerFlowable);
-                  CompletableFuture<R> future = stage.toCompletableFuture();
-                  // This will always be true unless there is a store
-                  if (future.isDone()) {
-                     if (listener.segmentsLost.contains(segment)) {
-                        return CompletableFutures.<R>completedNull();
-                     }
-                     return future;
-                  }
+   protected <R> CompletionStage<PublisherResult<R>> exactlyOnceHandleLostSegments(CompletionStage<R> finalValue, SegmentListener listener) {
+      return handleLostSegments(finalValue, listener);
+   }
 
-                  return stage.thenCompose(value -> {
-                     // This means the segment was lost in the middle of processing
-                     if (listener.segmentsLost.contains(segment)) {
-                        return CompletableFutures.completedNull();
-                     }
-                     return CompletableFuture.completedFuture(value);
-                  });
-               }).sequential();
+   /**
+    * This method iteratively submits a task to operate on the cpu bound thread pool for each segment that is available.
+    * The tasks perform a type of work stealing where they attempt to retrieve the next available segment and process
+    * them as fast as possible. It is entirely possible that a given task is never submitted due to the other tasks
+    * completing all the segments asynchronously. After the main thread has submitted all the tasks it will attempt
+    * to steal a segment and run it if possible and if it can will subsequently attempt to complete all remaining
+    * segments in the same fashion as the other threads. NOTE that this behavior is not normally how reactive
+    * streams are done as given operations are not normally performed until the returned Flowable is subscribed to, but
+    * for performance reasons this method eagerly publishes entries.
+    * <p>
+    * The results of each segment data will then be published each as a single result in the returned Flowable. Due
+    * to the results being retrieved eagerly it is entirely possible that if the Subscriber of the Flowable is slow
+    * that that results queue up. But due to that the fact that results are reduced to single values for each segment
+    * this shouldn't become an issue.
+    * @param set CacheSet to retrieve the publisher for (non-nullable)
+    * @param keysToExclude whether given keys should be excluded from the processing (nullable)
+    * @param toKeyFunction function to convert an entry to a key to determine if it is excluded (must be non null if keysToExclude is)
+    * @param segments the segments to process results for (non-nullable)
+    * @param transformer reducer to collate all the entries for a given segment into a single result (non-nullable)
```

Comment:
It runs things like the "reducers" in https://github.com/infinispan/infinispan/blob/master/core/src/main/java/org/infinispan/reactive/publisher/PublisherReducers.java.

It can be any user supplied one, but this is not exposed currently as I mentioned in the other comment. Currently it is all CPU bound operations that eventually return a single value. You can think of it like a local reduction for a segment worth of data (in most cases).

Reply:
I can rename the variable though.